### ISSUE TITLE
Strange behavior of login with authentication bypass 

### ISSUE THREAD

**Issue by grasshide** [VueTorrent/VueTorrent#1350]
I use the hotio docker of qbt with the vue theme. It is behind traefik, protected by the Authelia SSO portal.
In settings/webui I selected "Bypass authentication for clients whitelisted IP subnets" and put in all private networks:
- 10.0.0.0/8
- 172.16.0.0/12
- 192.168.0.0/16

The goal was that qbt would not ask for a password at all, since the traffic always comes from a private network via traefik. And the UI is protected by Authelia.

This led to a very inconsistent experience.

- The first login on a fresh browser (nothing in cache/ no cookies) works always fine
- Every second visit of the page has mixed outcomes:
  -  Almost always I can see the sidebar (with actual stats). Sometimes with torrents in the main window, sometimes with the login mask. The menu on top is also visible. I can click on "log" and close it again and I see my torrent list. Putting in my credentials to the login mask usually leads to an error. Refreshing the page sometimes also removes the login and shows the torrents.

**Comment by Larsluph** [comment]
Authentication flow really need to be reworked. It's not the first time we have some issues like that.

If you see the sidebar / topbar and data isn't zeroed, it means that you are logged in and that we can communicate with qbit. Automatic redirection might not work in some cases which displays the login form instead of the dashboard.

**Comment by jsaltzman** [comment]
I am having same issue. I have login turned off because I am behind authentik. When I load page I'm presented with a login box and if I type any user/pass and submit I get in. Sidebar shows with real data so I am clearly already logged in.

**Comment by ruifung** [comment]
Im pretty sure its the service worker caching the UI itself.

Proxy auth requires the browser to redirect to the login page, but unfortunately the serviceworker makes it load the UI from the cache while the qBt api calls just keep failing with 401 status codes in the background.

**Comment by Larsluph** [comment]
Thanks a lot for this investigation! 🙏

Unfortunately the service worker is required to make the PWA work. It is configurable so we could see what is really causing that bug 🐛

**Comment by ruifung** [comment]
I think the biggest problem is that the initial page load (html) is cached by the serviceworker, and thus the redirect from the proxy isn't even happening because it never fetched it at all.
Since this won't work if it can't access the server (obviously), maybe just exclude the main HTML that the browser loads for everything else?

**Comment by Larsluph** [comment]
The `index.html` only tells the browser how to start the app. Everything else is done in the main JS script which shouldn't be cached. This is really weird 🤔

**Comment by ruifung** [comment]
That's the whole issue no? If the index.html is cached, it means the browser doesn't attempt to load it from the source.

Because the browser doesn't load it from the source, it can't be redirected to another page to handle proxy auth.

Because it didn't get authenticated via proxy auth, the JS that was loaded by the serviceworker from it's cache just keep trying to make API calls that is blocked by proxy.

Solution: Never cache index.html, and make sure redirects from the server are properly handled and passed to the actual page.

Basically, the authentication by proxy, should happen *before* the JS is even started.

**Comment by Larsluph** [comment]
I explained myself badly earlier.

The only point of the html is to load the JS, caching the html would only cause the JS to not load if it has changed and the cache wasn't updated. JS isn't cached, at least that's how it should be.

The API calls aren't blocked because in that case all addresses are whitelisted so we can make API calls (that's why the sidebar displays correct information)

Every logic, redirect and stuff is done by the JS. Not sure how I can check for these cases as I'm not able to reproduce it.
You should be able to see the external calls made by your browser in the DevTools and when cache is used.

**Comment by ruifung** [comment]
I'm looking at the sw.js code from the docker image I'm running, I can see that the index.html is indeed precached by the serviceworker. Which has a revision attached to it in the precache call.

And as per the documentation for workbox, this means that all the files are indeed precached and loaded at serviceworker install time. Which is not the same as PWA install since the serviceworker is defined and activated on page load.

```
s.precacheAndRoute([{
        url: "apple-touch-icon.png",
        revision: "eed821a5bbfee3a20aab9c51c4977743"
    }, {
        url: "assets/ArrayPagination-JjYmrbrK.js",
        revision: null
    }, {
        url: "assets/Dashboard-fTAYIqJw.css",
        revision: null
    }, {
        url: "assets/Dashboard-VdGBTYWl.js",
        revision: null
    }, {
        url: "assets/dayjs-PZR0PBzP.js",
        revision: null
    }, {
        url: "assets/dayjs.min-lc5slf-X.js",
        revision: null
    }, {
        url: "assets/index-fj5-vamU.js",
        revision: null
    }, {
        url: "assets/index-sIPFeh3n.js",
        revision: null
    }, {
        url: "assets/index-XtlGFseb.css",
        revision: null
    }, {
        url: "assets/Login-9Kk998Ee.js",
        revision: null
    }, {
        url: "assets/Logs-61CqD2Lh.css",
        revision: null
    }, {
        url: "assets/Logs-tlAtjJPZ.js",
        revision: null
    }, {
        url: "assets/MagnetHandler-nMflNbET.js",
        revision: null
    }, {
        url: "assets/materialdesignicons-webfont-6eb_lmTU.woff2",
        revision: null
    }, {
        url: "assets/materialdesignicons-webfont-D15t_tsC.woff",
        revision: null
    }, {
        url: "assets/PasswordField.vue_vue_type_script_setup_true_lang-3N6bRpUb.js",
        revision: null
    }, {
        url: "assets/roboto-cyrillic-400-normal-1Q02bZlk.woff2",
        revision: null
    }, {
        url: "assets/roboto-cyrillic-400-normal-wkKjpXzZ.woff",
        revision: null
    }, {
        url: "assets/roboto-cyrillic-ext-400-normal-PiqLoFV_.woff",
        revision: null
    }, {
        url: "assets/roboto-cyrillic-ext-400-normal-zkSvWxgI.woff2",
        revision: null
    }, {
        url: "assets/roboto-greek-400-normal-UVhwlGKP.woff2",
        revision: null
    }, {
        url: "assets/roboto-greek-400-normal-ZxjWinlq.woff",
        revision: null
    }, {
        url: "assets/roboto-latin-400-normal-JkyEVz-m.woff2",
        revision: null
    }, {
        url: "assets/roboto-latin-400-normal-VNUqCuId.woff",
        revision: null
    }, {
        url: "assets/roboto-latin-ext-400-normal-5aATcKHE.woff",
        revision: null
    }, {
        url: "assets/roboto-latin-ext-400-normal-OGy6Zcg4.woff2",
        revision: null
    }, {
        url: "assets/roboto-mono-cyrillic-400-normal-eoOUqCpM.woff",
        revision: null
    }, {
        url: "assets/roboto-mono-cyrillic-400-normal-xDevZbqQ.woff2",
        revision: null
    }, {
        url: "assets/roboto-mono-cyrillic-ext-400-normal-EOqNCTb0.woff2",
        revision: null
    }, {
        url: "assets/roboto-mono-cyrillic-ext-400-normal-uLllddJE.woff",
        revision: null
    }, {
        url: "assets/roboto-mono-greek-400-normal-6dB6LRr6.woff",
        revision: null
    }, {
        url: "assets/roboto-mono-greek-400-normal-XRETARFo.woff2",
        revision: null
    }, {
        url: "assets/roboto-mono-latin-400-normal-dLuruHlQ.woff",
        revision: null
    }, {
        url: "assets/roboto-mono-latin-400-normal-v-cFAluQ.woff2",
        revision: null
    }, {
        url: "assets/roboto-mono-latin-ext-400-normal-_xD1OgjO.woff2",
        revision: null
    }, {
        url: "assets/roboto-mono-latin-ext-400-normal-hZ4_iJW8.woff",
        revision: null
    }, {
        url: "assets/roboto-mono-vietnamese-400-normal-SW7adlF6.woff",
        revision: null
    }, {
        url: "assets/roboto-mono-vietnamese-400-normal-SZ_VqEuC.woff2",
        revision: null
    }, {
        url: "assets/roboto-vietnamese-400-normal-JAkXt1WZ.woff2",
        revision: null
    }, {
        url: "assets/roboto-vietnamese-400-normal-ZBATgFfY.woff",
        revision: null
    }, {
        url: "assets/RssArticles-_OulcYvY.css",
        revision: null
    }, {
        url: "assets/RssArticles-bvG5sT4p.js",
        revision: null
    }, {
        url: "assets/SearchEngine-DBKkjdr0.js",
        revision: null
    }, {
        url: "assets/Settings-HpcqpWK7.css",
        revision: null
    }, {
        url: "assets/Settings-kYE7F9dE.js",
        revision: null
    }, {
        url: "assets/TorrentDetail-eWfy6OX1.css",
        revision: null
    }, {
        url: "assets/TorrentDetail-L691_dpm.js",
        revision: null
    }, {
        url: "assets/vue-EkRptf2M.js",
        revision: null
    }, {
        url: "assets/vuetify-FD_HUlNY.js",
        revision: null
    }, {
        url: "favicon.ico",
        revision: "621ef878b9a700bd2249dad60d9f1141"
    }, {
        url: "icon-192.png",
        revision: "1a475d65497c0b3609661246d8639ff4"
    }, {
        url: "icon-512.png",
        revision: "fa311af470073dfeb0673806ae3e0e2f"
    }, {
        url: "icon.svg",
        revision: "1a5efa1aa226aa0533605b7e84667ecd"
    }, {
        url: "index.html",
        revision: "b1fb2f21498570affbdd26a6c86178f1"
    }, {
        url: "registerSW.js",
        revision: "402b66900e731ca748771b6fc5e7a068"
    }, {
        url: "favicon.ico",
        revision: "621ef878b9a700bd2249dad60d9f1141"
    }, {
        url: "icon.svg",
        revision: "1a5efa1aa226aa0533605b7e84667ecd"
    }, {
        url: "icon-192.png",
        revision: "1a475d65497c0b3609661246d8639ff4"
    }, {
        url: "icon-512.png",
        revision: "fa311af470073dfeb0673806ae3e0e2f"
    }, {
        url: "robots.txt",
        revision: "b6216d61c03e6ce0c9aea6ca7808f7ca"
    }, {
        url: "manifest.webmanifest",
        revision: "112d2a686dfcb44742c79ee217315d5c"
    }
```

**Comment by ruifung** [comment]
In addition, most of the JS and index.html is indeed cached as can be seen from this screenshot where the webUI was attempted to be loaded with the service stopped (which behaves identically to the proxy scenario). As you can see all the things are being loaded from the serviceworker cache.

![image](https://github.com/VueTorrent/VueTorrent/assets/1073945/bf067a8b-b01c-4e38-ab0a-65440788e1fe)

**Comment by ruifung** [comment]
Also I'd like to point out that in my experience running it with the serviceworker blocked, I have not noticed any issues with navigating around the UI. Since it's using URL Fragment to hold the navigation path.

So I'm actually curious to know what part of being a PWA will be broken just by not having a service worker?

**Comment by Larsluph** [comment]
> So I'm actually curious to know what part of being a PWA will be broken just by not having a service worker?

Service worker is a mandatory component of PWA, it can be a dummy one but it must exist for the PWA to work.

You said it works by only disabling cache on the `index.html`?

The weird thing I don't understand is how caching **static files** can break app behavior, while API calls are still being made...

**Comment by ruifung** [comment]
That's because you're thinking purely in the context of the app. This involves server side setup and usually reverse proxies which act between the app and qBittorrent, and need to be able to perform redirects BEFORE the app is loaded. (Hence, not caching index.html)

For example, in my setup, I have authentication basically disabled in qBittorrent, and enforce access via the reverse proxy and network policies ( I have it deployed in a kubernetes cluster locally )

The issue at hand here, is that we're adding an authentication layer between qbittorrent and the app via a reverse proxy (i.e. Oauth-proxy, authentik, etc.)

This layer blocks all access to API and typically maintains authentication state by a cookie set by the authentication flow.

Caching the static assets, means the browser doesn't attempt to load it from the server, and thus the reverse proxy doesn't get a chance to redirect, and thus no valid session cookie is set because no external authentication is performed, and thus ALL api calls just silently fail because they are not authenticated by the reverse proxy.

Disabling cache on the index.html means that when the browser loads the page it's something like this:
(Using an authentik setup as an example)

On first load or with index.html not cached.
1. Browser attempts to load index.html
2. Reverse proxy sees it has no valid session, issues redirect to external login page / service.
3. External login service redirects back to the qbittorrent domain
4. Reverse proxy sees a valid session and allows the request.
5. Index.html is loaded and bootstraps app
6. Everything proceeds as usual.

With the index.html precached, the following happens:
1. Browser attempts to load index.html
2. Service worker serves it from cache
3. Index.html is loaded and bootstraps app
4. App attempts API requests.
5. API requests silently fail due to reverse proxy blocking.
6. App displays login screen because it thinks it needs login.
7. Login screen will never work because it assumes qBittorrent authentication.
8. User resolves it by wiping site data which resets serviceworker cache, and thus triggering the 1st flow as shown above on the next load.

So my question is, given that this requires access to a qBittorrent instance, why even bother caching any static assets at all? I'll also note this somewhat breaks switching back to the default qBittorrent webUI without clearing the browser cache.

**Comment by Larsluph** [comment]
@ruifung I created [a branch](https://github.com/VueTorrent/VueTorrent/tree/fix/sw) to exclude the `index.html` from the service worker's pre-cache.

Could you try building this one to test if it works? I'm not able to test it as it doesn't seems to cache it on my setup...

**Comment by ruifung** [comment]
@Larsluph I mean, sure I guess. I just have to download everything required to set up a vue development environment to build it first. Maybe on the weekend I guess.

**Comment by Larsluph** [comment]
Here's the build

[vuetorrent.zip](https://github.com/VueTorrent/VueTorrent/files/13978807/vuetorrent.zip)

**Comment by ruifung** [comment]
Seems like the generated sw.js has a call to s.createHandlerBoundToURL("index.html") which is setting up a explicit route handler using the precached index.html

```
s.registerRoute(new s.NavigationRoute(s.createHandlerBoundToURL("index.html")))
```


I'm not sure if the workbox generateSW function can generate an alternate behavior. And while I recognize what it's trying to do (as in route everything to index.html, as per PWA navigation style), I'm not entirely sure if it's entirely required. Since it errored out, and yet vuetorrent is working just fine on my qBt instance.

```
Uncaught (in promise) non-precached-url: non-precached-url :: [{"url":"index.html"}]
    at O.createHandlerBoundToURL (https://qbittorrent.home.arpa/workbox-3e911b1d.js:1:13245)
    at Object.createHandlerBoundToURL (https://qbittorrent.home.arpa/workbox-3e911b1d.js:1:14916)
    at https://qbittorrent.home.arpa/sw.js:1:4599
    at https://qbittorrent.home.arpa/sw.js:1:565
createHandlerBoundToURL @ workbox-3e911b1d.js:1
(anonymous) @ workbox-3e911b1d.js:1
(anonymous) @ sw.js:1
(anonymous) @ sw.js:1
Promise.then (async)
(anonymous) @ sw.js:1
(anonymous) @ sw.js:1
```

**Comment by Larsluph** [comment]
I updated the PR to completely remove caching in the service worker.
